### PR TITLE
Fix systemd_unit user context

### PR DESCRIPTION
**Pull request by mal**
### Description

- Switches out `Etc.getpwuid` for `Etc.getpwnam` in `systemd_unit` provider and `systemd` provider for `service` resource in order to correctly handle usernames set on the resources.
- Fixes issue where `daemon-reload` would be called at the system level for user level unit changes which would leave the updated user units in an incomplete state.

### Issues Resolved

https://github.com/chef/chef/issues/7269

### Check List

- [x] New functionality includes tests
- [x] All tests pass
- [x] RELEASE\_NOTES.md, has been updated if required (not required for bugfixes, required for API changes)
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: lib/chef/provider/service/systemd.rb [hunk 1]
```diff
@@ -77,7 +77,7 @@ def define_resource_requirements
 
   def get_systemctl_options_args
     if new_resource.user
-      uid = Etc.getpwuid(new_resource.user).uid
+      uid = Etc.getpwnam(new_resource.user).uid
```
Comment (thommay) [inline lib/chef/provider/service/systemd.rb]: This'll explode if users have previously been using UIDs here:
```
irb(main):003:0> Etc.getpwnam 501
Traceback (most recent call last):
        3: from /Users/thom/.rubies/ruby-2.5.1/bin/irb:11:in `<main>'
        2: from (irb):3
        1: from (irb):3:in `getpwnam'
TypeError (no implicit conversion of Integer into String)
```
probably better to call the right thing if we have a string or an int.

**Comment by mal** [comment]
I don't understand why the tests fail, it runs fine on my local machine. The errors read like the mocking of `Etc.getpwnam` isn't being done correctly, but it looks correct. Any pointers would be very welcome.

File: lib/chef/provider/service/systemd.rb (hunk 1 above)
Comment (lamont-granquist) [reply lib/chef/provider/service/systemd.rb]: regardless we should accept both in the API.

there's fairly horrible problems with glibc caching+ruby and when chef creates the username where uids are required in apis like this in order to work around the problems (or else people need to run nscd, but everyone irrationally hates nscd).

Comment (mal) [reply lib/chef/provider/service/systemd.rb]: Any hope of that being a separate task? This was just meant to be a quick fix for something that's currently outright broken. Adding support and accompanying tests for allowing `user` to be a UID in the `service` and `systemd_unit` resources is not something I really want to take on.

Comment (thommay) [reply lib/chef/provider/service/systemd.rb]: Yeah, let's file a bug for that and merge this.

**Comment by thommay** [comment]
Thanks @mal, this is a great fix!

**Pull request merged**